app/lib/actions/bulk_action.rb

```diff
@@ -72,5 +73,11 @@ def batch(from, size)
     def total_count
       input[:target_ids].count
     end
+
+    private
+
+    def extract_concurrency_limit(args = [], limit = nil)
+      args.find { |arg| arg.is_a?(Hash) && arg.key?(:concurrency_limit) }&.fetch(:concurrency_limit, limit)
```

Comment:
My Ruby is bad, but doesn't this mean:
1. take `args`
2. find the first element in it that is a has and has a `concurrency_limit` key
3. if found, fetch the value. if not found, don't do anything as `find` returned `nil` and the safe operator skips the `.fetch`

so instead, how about:
```suggestion
      args.find { |arg| arg.is_a?(Hash) && arg.key?(:concurrency_limit) }&.fetch(:concurrency_limit) || limit
```

Reply:
(and depending on the Ruby version, calling with `(…, concurrency_limit: 25)` places that in either `args` or `concurrency_limit` and produces 2.7 failures